packages/devtools_app/lib/src/shared/diagnostics/inspector_service.dart

```diff
@@ -483,60 +485,69 @@ class InspectorService extends InspectorServiceBase {
     }
-    // TODO(jacobr): it would be nice to use Isolate.rootLib similar to how
-    // debugger.dart does but we are currently blocked by the
-    // --track-widget-creation transformer generating absolute paths instead of
-    // package:paths.
-    // Once https://github.com/flutter/flutter/issues/26615 is fixed we will be
-    // able to use package: paths. Temporarily all tools tracking widget
-    // locations will need to support both path formats.
+    // TODO(jacobr): Once https://github.com/flutter/flutter/issues/26615 is
+    // fixed we will be able to use package: paths. Temporarily all tools
+    // tracking widget locations will need to support both path formats.
     // TODO(jacobr): use the list of loaded scripts to determine the appropriate
     // package root directory given that the root script of this project is in
     // this directory rather than guessing based on url structure.
     final parts = path.split('/');
     String? pubRootDirectory;
-    for (int i = parts.length - 1; i >= 0; i--) {
-      final part = parts[i];
-      if (part == 'lib' || part == 'web') {
-        pubRootDirectory = parts.sublist(0, i).join('/');
-        break;
-      }
+    // For google3, we grab the top-level directory in the google3 directory
+    // (e.g. /education), or the top-level directory in third_party (e.g.
+    // /third_party/dart):
+    if (_isGoogle3Path(parts)) {
+      pubRootDirectory = _pubRootDirectoryForGoogle3(parts);
+    } else {
+      final parts = path.split('/');
 
-      if (part == 'packages') {
-        pubRootDirectory = parts.sublist(0, i + 1).join('/');
-        break;
+      for (int i = parts.length - 1; i >= 0; i--) {
+        final part = parts[i];
+        if (part == 'lib' || part == 'web') {
+          pubRootDirectory = parts.sublist(0, i).join('/');
+          break;
+        }
+
+        if (part == 'packages') {
+          pubRootDirectory = parts.sublist(0, i + 1).join('/');
+          break;
+        }
       }
     }
     pubRootDirectory ??= (parts..removeLast()).join('/');
 
     await _addPubRootDirectories([pubRootDirectory]);
-    await group.dispose();
     return pubRootDirectory;
   }
 
+  bool _isGoogle3Path(List<String> pathParts) =>
+      pathParts.contains(_google3PathSegment);
+
+  List<String> _stripGoogle3(List<String> pathParts) {
+    final google3Index = pathParts.lastIndexOf(_google3PathSegment);
+    if (google3Index != -1 && google3Index + 1 < pathParts.length) {
+      return pathParts.sublist(google3Index + 1);
+    }
+    return pathParts;
+  }
+
+  String? _pubRootDirectoryForGoogle3(List<String> pathParts) {
+    final strippedParts = _stripGoogle3(pathParts);
+    if (strippedParts.isEmpty) return null;
+
+    final topLevelDirectory = strippedParts.first;
+    if (topLevelDirectory == _thirdPartyPathSegment &&
+        strippedParts.length >= 2) {
+      return '/${strippedParts.sublist(0, 2).join('/')}';
+    } else {
+      return '/${strippedParts.first}';
```

Comment:
do we need to be concerned with how these paths will work with Windows at all? WRT forward slashes vs back slashes?

Reply:
I agree we should use package:path to be safe. I'll leave that up to you whether you want to do that here or file an issue to track this as a follow up. LGTM from me.

Reply:
SG, will do as follow up shortly (tracking in https://github.com/flutter/devtools/issues/6868)